File: ios/ClawChatKit/Sources/ClawChatKit/Protocol/ErrorMessages.swift
import Foundation

// MARK: - Error Code

public enum ClawChatErrorCode: Codable, Sendable, Equatable {
    case gatewayOffline
    case agentNotFound
    case sessionNotFound
    case unauthorized
    case rateLimited
    case payloadTooLarge
    case internalError
    case invalidMessage
    case incompatibleVersion
    case unknown(String)

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        switch raw {
        case "gateway_offline": self = .gatewayOffline
        case "agent_not_found": self = .agentNotFound
        case "session_not_found": self = .sessionNotFound
        case "unauthorized": self = .unauthorized
        case "rate_limited": self = .rateLimited
        case "payload_too_large": self = .payloadTooLarge
        case "internal_error": self = .internalError
        case "invalid_message": self = .invalidMessage
        case "incompatible_version": self = .incompatibleVersion
        default: self = .unknown(raw)
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .gatewayOffline: try container.encode("gateway_offline")
        case .agentNotFound: try container.encode("agent_not_found")
        case .sessionNotFound: try container.encode("session_not_found")
        case .unauthorized: try container.encode("unauthorized")
        case .rateLimited: try container.encode("rate_limited")
        case .payloadTooLarge: try container.encode("payload_too_large")
        case .internalError: try container.encode("internal_error")
        case .invalidMessage: try container.encode("invalid_message")
        case .incompatibleVersion: try container.encode("incompatible_version")
        case .unknown(let raw): try container.encode(raw)
        }
    }
}

// MARK: - Error Message

public struct ErrorMessage: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
    public let code: ClawChatErrorCode
    public let message: String
    public let requestId: String?
}

File: android/clawchat-kit/src/main/kotlin/com/clawchat/kit/protocol/ErrorMessages.kt
package com.clawchat.kit.protocol

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

// MARK: - Error Code

@Serializable(with = ClawChatErrorCodeSerializer::class)
sealed class ClawChatErrorCode {
    data object GatewayOffline : ClawChatErrorCode()
    data object Unauthorized : ClawChatErrorCode()
    data object NotConnected : ClawChatErrorCode()
    data object InvalidJson : ClawChatErrorCode()
    data object RateLimited : ClawChatErrorCode()
    data class Unknown(val raw: String) : ClawChatErrorCode()

    val value: String get() = when (this) {
        is GatewayOffline -> "gateway_offline"
        is Unauthorized -> "unauthorized"
        is NotConnected -> "not_connected"
        is InvalidJson -> "invalid_json"
        is RateLimited -> "rate_limited"
        is Unknown -> raw
    }

    companion object {
        fun fromString(s: String): ClawChatErrorCode = when (s) {
            "gateway_offline" -> GatewayOffline
            "unauthorized" -> Unauthorized
            "not_connected" -> NotConnected
            "invalid_json" -> InvalidJson
            "rate_limited" -> RateLimited
            else -> Unknown(s)
        }
    }
}

object ClawChatErrorCodeSerializer : KSerializer<ClawChatErrorCode> {
    override val descriptor = PrimitiveSerialDescriptor("ClawChatErrorCode", PrimitiveKind.STRING)
    override fun serialize(encoder: Encoder, value: ClawChatErrorCode) = encoder.encodeString(value.value)
    override fun deserialize(decoder: Decoder): ClawChatErrorCode = ClawChatErrorCode.fromString(decoder.decodeString())
}

// MARK: - Error Message

@Serializable
data class ErrorMessage(
    override val type: String,
    override val id: String,
    override val ts: Long,
    val code: ClawChatErrorCode,
    val message: String,
) : BaseMessage
